### PR TITLE
[FEA] Add zstd compression support for `to_csv`

### PR DESCRIPTION
**Pull request by a-hirota**
## Description

Adds GPU-accelerated ZSTD compression to the CSV writer, matching the pandas `compression="zstd"` API.

Why ZSTD only:
- nvCOMP provides GPU-accelerated ZSTD compression
- ZSTD supports concatenated frames, so each chunk can be compressed independently while the resulting file remains a single stream that standard tools (`zstd -d`) can decompress
- Other codecs are possible as follow-ups; this change focuses on ZSTD

Implementation:
- Add `compression` to `csv_writer_options` and its builder, rejecting codecs other than `NONE`/`ZSTD` at option-construction time
- Compress via the existing `cudf::io::detail::compress` API, so the writer inherits the shared host/device dispatch rather than re-implementing nvCOMP setup
- Fail loudly if compression fails, instead of silently writing uncompressed bytes into a `.zst` file
- Plumb `compression` through the pylibcudf bindings and `cudf.to_csv`, rejecting the combination with a string return value since compressed output is binary

Usage:
```python
df.to_csv("output.csv.zst", compression="zstd", chunksize=10000)
```

## Notes

Taken over from @a-hirota (original authorship preserved). Rebased onto current `main` and the outstanding review items are addressed; see the comment below for details.

Known follow-ups, deliberately out of scope here so the feature can land and be measured first:
- Decouple `rows_per_chunk` from the ZSTD frame size (raised by @vuule)
- Fold the header and trailing newline into the adjacent data frame instead of emitting them as their own frames
- Reuse the compression output buffer across chunks rather than allocating per write

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/csv/writer_impl.cu [hunk 1]
```diff
@@ -372,23 +535,21 @@ void write_chunked(data_sink* out_sink,
...
-                         options.get_line_terminator().size());
+  // Needs newline at the end, to separate from next chunk (only for uncompressed writes)
+  if (compression == compression_type::NONE) {
+    if (out_sink->is_device_write_preferred(newline.size())) {
+      out_sink->device_write(newline.data(), newline.size(), stream);
+    } else {
+      out_sink->host_write(options.get_line_terminator().data(),
+                           options.get_line_terminator().size());
+    }
   }
```
[the remark this answers is not in the document]
Comment (a-hirota) [reply cpp/src/io/csv/writer_impl.cu]: Fixed. Writing newline as separate ZSTD frame for memory efficiency.

File: cpp/tests/io/csv_test.cpp [hunk 2]
```diff
@@ -2782,4 +2782,62 @@ TEST_F(CsvReaderTest, CommentLinesWithQuotedStrings)
...
+    cudf::io::csv_writer_options::builder(cudf::io::sink_info(&compressed_buffer), input_table)
+      .include_header(true)
+      .names({"value"})
+      .rows_per_chunk(10)  // Force multiple chunks
+      .compression(cudf::io::compression_type::ZSTD);
+  cudf::io::write_csv(opts);
+
+  // Verify output was produced
+  EXPECT_GT(compressed_buffer.size(), 0);
+}
```
[the remark this answers is not in the document]
Comment (a-hirota) [reply cpp/tests/io/csv_test.cpp]: Fixed. Added content verification test.

File: cpp/include/cudf/io/csv.hpp [hunk 3]
```diff
@@ -1583,6 +1592,17 @@ class csv_writer_options {
...
+  /**
+   * @brief Sets the compression type for the output.
+   *
+   * Only ZSTD compression is supported for CSV writer because it allows
+   * concatenated frames, enabling progressive chunk-based compression
+   * compatible with standard decompression tools.
+   *
+   * @param comp The compression type (NONE or ZSTD only)
+   */
+  void set_compression(compression_type comp) { _compression = comp; }
```
[the remark this answers is not in the document]
Comment (a-hirota) [reply cpp/include/cudf/io/csv.hpp]: Fixed. Added CUDF_EXPECTS validation.

File: python/pylibcudf/pylibcudf/io/csv.pyx [hunk 4]
```diff
@@ -869,6 +869,22 @@ cdef class CsvWriterOptionsBuilder:
...
+        comp : CompressionType
+            Compression type (NONE, ZSTD, etc.)
+
+        Returns
+        -------
+        CsvWriterOptionsBuilder
+            Builder to build CsvWriterOptions
+        """
+        self.c_obj.compression(comp)
+        return self
```
[the remark this answers is not in the document]
Comment (a-hirota) [reply python/pylibcudf/pylibcudf/io/csv.pyx:913]: Fixed. Updated docstring.

File: python/cudf/cudf/io/csv.py [hunk 5]
```diff
@@ -493,6 +507,7 @@ def _plc_write_csv(
                 else (str(name) if name not in (None, "") else "")
                 for name in all_names
             ]
+
```
[the remark this answers is not in the document]
Comment (a-hirota) [reply python/cudf/cudf/io/csv.py]: Fixed.

**Comment by mhaseeb123** [comment]
Pasting the review by Claude here (still manually validating if they make sense). Please exercise caution:

<details>
  <summary>Click to expand</summary>

## Findings

### Critical

- **[writer_impl.cu, `compress_chunk`, ~lines 100-105]** `thrust::copy` with
  `rmm::exec_policy_nosync(stream)` is used to copy **host-resident** `device_span` structs
  (`input_span`, `output_span`) into device uvectors. The source pointers (`&input_span`,
  `&input_span + 1`) point to stack-local host memory, but `rmm::exec_policy_nosync` uses a
  **device** execution policy — this is a device-to-device Thrust call being fed host pointers,
  which is **undefined behavior**. The spans are trivially-copyable structs living on the host
  stack. Use `cudaMemcpyAsync` (host-to-device) or use `cudf::detail::hostdevice_vector` as the
  existing compression infrastructure does. Compare with how `device_compress` in
  `compression.cpp` receives its inputs: callers use `hostdevice_vector` and call
  `host_to_device_async` to transfer spans to the device before passing them as `device_span`.

- **[writer_impl.cu, `compress_chunk`, ~line 108]** Same issue with `thrust::fill` on `results` —
  `init_result` is a host-local struct being used as the fill value in a device execution policy.
  While Thrust may handle scalar fill values from host memory in some backends, this is fragile
  and inconsistent with the rest of the codebase.

- **[writer_impl.cu, `compress_chunk`]** The function should **not** duplicate the batched
  compression setup that already exists in `device_compress()` (`compression.cpp`). The existing
  `device_compress` function handles nvCOMP dispatch, disabled-check, and fallback. The new code
  re-implements all of this (nvcomp type mapping, `is_compression_disabled` check, batched
  compress call, result checking) — this is significant code duplication and diverges from the
  established pattern. Instead, use `device_compress` directly, which already accepts
  `device_span<device_span<uint8_t const> const>` inputs/outputs/results. This eliminates ~80
  lines of error-prone code.

- **[writer_impl.cu, `write_data_with_compression`]** The `write_data_with_compression` function
  allocates a fresh `rmm::device_buffer` for the compressed output on **every call**. For chunked
  writes, this means N+1 allocations (N chunks + header). Since the max compressed size is
  predictable, the compression buffer should be allocated once and reused across chunks.

- **[writer_impl.cu, `write_chunked_begin`]** When compression is enabled, the header is copied
  to the GPU via `cudaMemcpyAsync`, compressed, then written. But the header is typically a few
  hundred bytes. Compressing such a tiny payload with ZSTD is wasteful (the compressed frame
  header overhead alone may exceed the data size). More importantly, the ZSTD frame for the header
  will be a separate frame from the data frames, which means decompression tools must handle
  concatenated frames starting from the very first frame. While this works with `zstd -d` and
  streaming decompressors, it adds unnecessary complexity. Consider writing the header uncompressed
  (it's plain text and tiny) and only compressing the data chunks, or concatenating the header
  into the first data chunk before compression.

- **[csv.py, `to_csv`, ~line 387-390]** When `path_or_buf is None`, the code creates a `StringIO`
  and sets `return_as_string = True`. But if `compression="zstd"` is also set, the compressed
  binary output will be written to a `StringIO` (text mode), which will fail or produce garbage.
  There is no guard against `compression` + `return_as_string`. Either raise an error for this
  combination or use `BytesIO` when compression is enabled.

### Suggestions

- **[writer_impl.cu, `compress_chunk`]** The silent fallback to uncompressed output when
  compression fails (returning 0) is dangerous for correctness. If the user explicitly requested
  ZSTD compression, silently writing uncompressed data means the output file will not be a valid
  ZSTD file, but the user won't know. This should either raise an error or at minimum log a
  warning. The `is_compression_disabled` check should also throw rather than silently fall back —
  if the user asked for ZSTD and nvCOMP is disabled via `LIBCUDF_NVCOMP_POLICY`, that's an error
  condition, not a "try uncompressed" condition.

- **[writer_impl.cu, `write_data_with_compression`]** The function has three near-identical
  code paths for writing to the sink (`device_write` vs `host_write` with
  `is_device_write_preferred`). This pattern already exists in the original code and is duplicated
  three times in the new function (compressed success, compressed failure fallback, uncompressed).
  Extract a small `write_to_sink(data_sink*, void const*, size_t, stream)` helper to avoid this
  3x duplication.

- **[csv.py, `_plc_write_csv`]** The compression string-to-enum mapping is done inline with
  `if compression == "zstd"` / `else`. This means any string other than `"zstd"` (including
  typos like `"Zstd"`) silently maps to `NONE`. The validation in `to_csv` catches unknown
  strings, but `_plc_write_csv` is a separate function that could be called independently. Add
  validation here too, or at least assert that `compression` is `None` or `"zstd"`.

- **[csv_test.cpp, `ZstdCompression`]** The first C++ test only checks that compressed output
  differs from uncompressed output. It does not verify that the compressed output can be
  decompressed back to the original data. The `ZstdCompressionChunked` test does this correctly
  by reading back via `read_csv` with ZSTD decompression. The basic test should do the same.

- **[test_csv.py]** The Python test `test_to_csv_zstd_compression` decompresses with
  `dctx.decompress(f.read())` which requires the entire content to fit in a single ZSTD frame.
  But the writer produces concatenated frames (header frame + data frame + newline frame). This
  may fail depending on the `zstandard` library version. The chunked test correctly uses
  `stream_reader` which handles concatenated frames. Both tests should use the streaming API for
  consistency and correctness.

- **[test_csv.py]** Missing edge-case tests: empty DataFrame with compression, single-row
  DataFrame, DataFrame with null values, DataFrame with only string columns (high compression
  ratio), and very large chunks that might stress the nvCOMP buffer sizing.

- **[csv.hpp, `set_compression`]** The setter is defined inline in the public header. This is
  fine for a simple setter, but it pulls in `CUDF_EXPECTS` (and transitively `<stdexcept>`) into
  the public header. Consider moving the validation to the `.cu` implementation file, keeping only
  the trivial assignment in the header, consistent with how `set_quoting` is implemented (which
  also validates inline, so this is at least consistent).

### Nits

- **[PR files]** The PR includes `.serena/cache/cpp/document_symbols.pkl` and
  `.serena/cache/cpp/raw_document_symbols.pkl` (binary pickle files) and a `PR_DESCRIPTION.md`
  file. These are clearly not intended to be part of the PR and should be removed. The
  `PR_DESCRIPTION.md` also describes a completely different feature (CP932/Shift-JIS encoding
  support) — this appears to be leftover from a different branch or tool.

- **[writer_impl.cu]** `using namespace cudf::io;` is already present at the top of the file.
  The new code uses `cudf::io::detail::nvcomp::` and `io::detail::codec_exec_result` with
  varying qualification levels. Be consistent with the existing namespace usage.

- **[writer_impl.cu, `compress_chunk`]** `nvcomp::compression_type nvcomp_type =
  nvcomp::compression_type::ZSTD;` — use `auto const` per cudf style, and this variable is
  redundant since the function already asserts ZSTD-only. Just use the literal directly.

- **[csv.hpp]** The `@return` tag says `compression_type The compression type for the output` —
  the type name in the description is redundant with the actual return type. Simplify to
  `@return The compression type for the output`.

</details>

**Comment by vuule** [comment]
Thank you for the PR, @a-hirota! I ran some benchmarks locally and it performs surprisingly well (given that we compress one block at a time). Would be nice to decouple chunk size from ZSDT block size, but this can be a separate PR.

Please let us know if you need help addressing the comments in the review that @mhaseeb123 has posted :)

**Comment by GregoryKimball** [comment]
@vuule Should we take this over and merge it?

**Comment by vuule** [comment]
Taking this over so it can land — thanks @a-hirota for the original work, authorship is preserved on the commit.

I rebased onto current `main` (the branch was ~1500 commits behind) and squashed to a single clean commit, dropping the stray `.serena/` cache files and `PR_DESCRIPTION.md`. Addressing @mhaseeb123's review:

- **Host pointers with a device execution policy.** Removed. The custom `compress_chunk` is gone; spans and results now go through `cudf::detail::hostdevice_vector` + `host_to_device_async`, matching the rest of the IO code. Worth noting the repo's own `exec-policy-nosync-memory-resource` pre-commit hook flags the old code, so this was blocking commits outright.
- **Duplicated nvCOMP setup.** Removed. The writer now calls `cudf::io::detail::compress`, so it inherits the shared host/device dispatch and the disabled-codec handling instead of re-deriving them. This also means a build with device ZSTD unavailable transparently uses the host path rather than needing writer-specific fallback logic.
- **Silent fallback to uncompressed output.** Fixed. A failed compression is now a `CUDF_EXPECTS` failure rather than emitting raw CSV bytes into a `.zst` file.
- **`set_compression` validation.** Kept and now covered by a test; invalid codecs throw at option construction, mirroring `set_quoting`.
- **Test coverage.** The C++ tests now round-trip through `read_csv` and compare against the *uncompressed* write read back the same way, which is what actually catches chunk-separator corruption. Added a no-header case and an invalid-codec case. The Python tests use the streaming decompressor (required for concatenated frames) and assert the decompressed bytes equal the uncompressed `to_csv` output exactly, across single-chunk, multi-chunk, single-row, and null-containing inputs.

One correctness hole not in the original review: `to_csv(compression="zstd")` with no `path_or_buf` wrote binary into a text-mode `StringIO`. That combination now raises, and `_plc_write_csv` no longer silently maps an unrecognized compression string to `NONE`.

**Comment by vyasr** [comment]
@vuule how do you want to deal with reviews here since you've taken over the PR?

**Comment by vuule** [comment]
> @vuule how do you want to deal with reviews here since you've taken over the PR?

yeah, just struggling to find time to continue working on this. Should come back to this within a week.

**Comment by vyasr** [comment]
> > @vuule how do you want to deal with reviews here since you've taken over the PR?
> 
> yeah, just struggling to find time to continue working on this. Should come back to this within a week.

No rush, just checking in before I review (or ping for reviews).

**Comment by vuule** [comment]
> No rush, just checking in before I review (or ping for reviews).

Not quite ready for review :)

File: cpp/src/io/functions.cpp [hunk 6]
```diff
@@ -67,7 +67,7 @@ compression_type infer_compression_type(compression_type compression, source_inf
   if (ext == "gz") { return compression_type::GZIP; }
   if (ext == "zip") { return compression_type::ZIP; }
   if (ext == "bz2") { return compression_type::BZIP2; }
-  if (ext == "zstd") { return compression_type::ZSTD; }
+  if (ext == "zst" or ext == "zstd") { return compression_type::ZSTD; }
```
Comment (vuule) [inline cpp/src/io/functions.cpp:70]: zst is standard, keeping the old extension to avoid any breakage.

File: python/cudf/cudf/io/csv.py [hunk 7]
```diff
@@ -397,6 +400,12 @@ def to_csv(
...
     return_as_string = False
     if path_or_buf is None:
+        # compressed output is binary, so it cannot be returned as a string
+        if compression:
+            raise ValueError(
+                f"Compression {compression} is not supported when returning "
+                "the CSV output as a string; please provide `path_or_buf`."
+            )
         path_or_buf = StringIO()
         return_as_string = True
```
Comment (mhaseeb123) [inline python/cudf/cudf/io/csv.py:469]: AI pointed out that the sink decodes host writes as UTF-8 for any `TextIOBase` target (`pylibcudf/io/types.pyx:653-656`), so compressed output still breaks for an explicit text sink, e.g. `df.to_csv(open(p, "w"), compression="zstd")`.

```suggestion
    return_as_string = False
    if path_or_buf is None:
        # compressed output is binary, so it cannot be returned as a string
        if compression:
            raise ValueError(
                f"Compression {compression} is not supported when returning "
                "the CSV output as a string; please provide `path_or_buf`."
            )
        path_or_buf = StringIO()
        return_as_string = True
    elif compression and isinstance(path_or_buf, TextIOBase):
        raise ValueError(
            f"Compression {compression} is not supported for text-mode "
            "sinks; please pass a path or a binary file object."
        )
```
(needs `TextIOBase` in the `io` import.)

Same problem for remote paths just below: `get_writer_filepath_or_buffer(..., mode="w")` opens fsspec files in text mode, so `df.to_csv("s3://b/f.csv.zst", compression="zstd")` fails with a `UnicodeDecodeError`. Passing `mode="wb" if compression else "w"` would fix it.

File: cpp/src/io/csv/writer_impl.cu [hunk 8]
```diff
@@ -57,6 +66,133 @@ using namespace cudf::io;
...
+                                                   std::min(block_size, data.size() - offset)};
+  }
+  if (not tail.empty()) {
+    h_inputs[num_blocks - 1] = device_span<uint8_t const>{as_bytes(tail), tail.size()};
+  }
+
+  // the output buffer is laid out using the maximum compressed size of each block, since the
+  // compressed sizes are not known yet, with padding to keep every block aligned for the codec
+  auto const alignment = io::detail::compress_required_chunk_alignment(compression);
+  auto out_offsets     = cudf::detail::make_host_vector<size_t>(num_blocks + 1, stream);
```
Comment (mhaseeb123) [inline cpp/src/io/csv/writer_impl.cu]: Very optional: `out_offsets` isn't copied to the device so we could just use `std::vector`

```suggestion
  auto out_offsets = std::vector<size_t>(num_blocks + 1);
```

File: cpp/src/io/csv/writer_impl.cu [hunk 9]
```diff
@@ -57,6 +66,133 @@ using namespace cudf::io;
...
+
+  io::detail::compress(compression, d_inputs, d_outputs, d_results, stream);
+
+  auto const results = cudf::detail::make_host_vector(d_results, stream);
+  CUDF_EXPECTS(std::all_of(results.begin(),
+                           results.end(),
+                           [](auto const& result) {
+                             return result.status == io::detail::codec_status::SUCCESS;
+                           }),
+               "Error in compression of CSV data");
```
Comment (mhaseeb123) [inline cpp/src/io/csv/writer_impl.cu:166]: Optional: Consider including the first failing block index and its status in the message if possible.

File: cpp/src/io/csv/writer_impl.cu [hunk 10]
```diff
@@ -57,6 +66,133 @@ using namespace cudf::io;
...
+    total_comp_size += results[block].bytes_written;
+  }
+  auto const d_sources = cudf::detail::make_device_uvector_async(h_sources, stream, temp_mr);
+  auto const d_dests   = cudf::detail::make_device_uvector_async(h_dests, stream, temp_mr);
+  auto const d_sizes   = cudf::detail::make_device_uvector_async(h_sizes, stream, temp_mr);
+  cudf::detail::batched_memcpy_async(
+    d_sources.begin(), d_dests.begin(), d_sizes.begin(), num_blocks, stream);
+
+  write_to_sink(
+    out_sink, {reinterpret_cast<char const*>(packed_buffer.data()), total_comp_size}, stream);
```
Comment (mhaseeb123) [inline cpp/src/io/csv/writer_impl.cu:193]: If peak memory is a concern, we could skip the packing pass and issue a `device_write` per frame when the sink prefers device writes, here or in a follow-up PR that reuses the buffers. Currently `comp_buffer` and `packed_buffer` are both live alongside the chunk incurring ~2x transient mem usage.

File: cpp/include/cudf/io/csv.hpp [hunk 11]
```diff
@@ -1720,6 +1774,34 @@ class csv_writer_options_builder {
...
   }
 
+  /**
+   * @brief Sets the compression type for the output.
+   *
+   * Only NONE and ZSTD are supported.
+   *
+   * @param comp The compression type (NONE or ZSTD only)
+   * @return this for chaining
+   */
```
Comment (mhaseeb123) [inline cpp/include/cudf/io/csv.hpp:1803]:
```suggestion
  /**
   * @brief Sets the compression type for the output.
   *
   * Only NONE and ZSTD are supported.
   *
   * @param comp The compression type (NONE or ZSTD only)
   * @return this for chaining
   * @throw cudf::logic_error if compression type is not NONE or ZSTD
   */
```

File: cpp/include/cudf/io/csv.hpp [hunk 12]
```diff
@@ -1720,6 +1774,34 @@ class csv_writer_options_builder {
...
+  }
+
+  /**
+   * @brief Sets the size of the blocks that the output is compressed in.
+   *
+   * Independent of `rows_per_chunk`, and capped at the codec's maximum input size.
+   *
+   * @param size The compression block size, in bytes
+   * @return this for chaining
+   */
```
Comment (mhaseeb123) [inline cpp/include/cudf/io/csv.hpp:1818]:
```suggestion
  /**
   * @brief Sets the size of the blocks that the output is compressed in.
   *
   * Independent of `rows_per_chunk`, and capped at the codec's maximum input size.
   *
   * @param size The compression block size, in bytes
   * @return this for chaining
   * @throw cudf::logic_error if the block size is zero
   */
```

File: cpp/tests/io/csv_test.cpp [hunk 13]
```diff
@@ -2945,4 +2945,164 @@ TEST_F(CsvReaderTest, CommentLinesWithQuotedStrings)
...
+  std::vector<char> buffer;
+  auto int_col = column_wrapper<int32_t>{1, 2, 3};
+  cudf::table_view input_table(std::vector<cudf::column_view>{int_col});
+
+  EXPECT_THROW(cudf::io::csv_writer_options::builder(cudf::io::sink_info(&buffer), input_table)
+                 .compression(cudf::io::compression_type::SNAPPY),
+               cudf::logic_error);
+}
+
+TEST_F(CsvWriterTest, ZstdCompressionBlockSize)
```
Comment (mhaseeb123) [inline cpp/tests/io/csv_test.cpp:3136]: This test is single-chunk, so multi-chunk x multi-block never runs. Consider adding a small `rows_per_chunk` to one of the block sizes, plus the all-null, non-ASCII and sliced-input cases the test guidelines ask for.

File: python/cudf/cudf/utils/ioutils.py [hunk 14]
```diff
@@ -1310,9 +1311,8 @@
 encoding : str, default 'utf-8'
     A string representing the encoding to use in the output file
     Only 'utf-8' is currently supported
-compression : str, None
-    A string representing the compression scheme to use in the output file
-    Compression while writing csv is not supported currently
+compression : {{'zstd', None}}, default None
+    A string representing the compression scheme to use in the output file.
```
Comment (mhaseeb123) [inline python/cudf/cudf/utils/ioutils.py:1315]: Worth mentioning that the codec is not inferred from the file name, unlike pandas

```suggestion
compression : {{'zstd', None}}, default None
    A string representing the compression scheme to use in the output file.
    Not inferred from the file name; ``'infer'`` is not supported.
```

File: python/cudf/cudf/io/csv.py [hunk 15]
```diff
@@ -487,6 +499,17 @@ def _plc_write_csv(
...
+    # Map compression string to pylibcudf enum
+    if compression is None:
+        plc_compression = plc.io.types.CompressionType.NONE
+    elif compression == "zstd":
+        plc_compression = plc.io.types.CompressionType.ZSTD
+    else:
+        raise NotImplementedError(
+            f"Compression {compression} is not supported. "
+            "Only 'zstd' is supported for CSV compression."
+        )
```
Comment (mhaseeb123) [inline python/cudf/cudf/io/csv.py]: `to_csv` is the only caller and already rejects anything other than `None`/`"zstd"`

```suggestion
    plc_compression = (
        plc.io.types.CompressionType.ZSTD
        if compression == "zstd"
        else plc.io.types.CompressionType.NONE
    )
```

**Review by mhaseeb123 — COMMENTED** [review]
Nice cleanup. Some minor comments. Looking good overall

File: cpp/src/io/csv/writer_impl.cu (hunk 10 above)
Comment (vuule) [reply cpp/src/io/csv/writer_impl.cu:193]: Tried this, and it does not reduce the peak memory. The peak is during compression, because of the temp memory.

File: cpp/src/io/csv/writer_impl.cu (hunk 9 above)
Comment (vuule) [reply cpp/src/io/csv/writer_impl.cu:166]: Current code is consistent with other places with this check. If we change this, it should be in a separate PR across all formats.

File: python/cudf/cudf/io/csv.py (hunk 7 above)
Comment (vuule) [reply python/cudf/cudf/io/csv.py:469]: `SinkInfo` unwraps a `TextIOBase` sink to its .buffer, so `open(p, "w")` and remote paths (`fsspec.open(..., mode="w"`) also gives a `TextIOWrapper`) already write valid .zst (no `mode="wb"` needed). Only `io.StringIO` breaks, since it goes straight to `iobase_data_sink` and the `UnicodeDecodeError` escapes as an opaque `SystemError`. Guarding on `TextIOBase` would reject the working cases, so I keyed the check on the missing binary buffer instead, and added tests for both: `StringIO` now raises, while text, binary and memory:// sinks round-trip.

**Review by mhaseeb123 — APPROVED** [review]

File: cpp/src/io/csv/writer_impl.cu [hunk 16]
```diff
@@ -379,17 +525,25 @@ void write_chunked(data_sink* out_sink,
...
+    // The trailing newline that separates this chunk from the next one is compressed as an extra
+    // block of the same batched call, so it costs neither a copy of the chunk nor a second write.
+    write_compressed_to_sink(
+      out_sink,
+      data,
+      device_span<char const>{newline.data(), static_cast<size_t>(newline.size())},
+      compression,
+      block_size,
+      stream);
+    return;
```
Comment (vyasr) [inline cpp/src/io/csv/writer_impl.cu]: Minor nit: in case we ever want the write functions to return e.g. the number of bytes written, this might be cleaner to write as `return write_compressed_to_sink...` now so that the return value just flows through.

File: cpp/src/io/csv/writer_impl.cu [hunk 17]
```diff
@@ -57,6 +66,133 @@ using namespace cudf::io;
...
+  auto const size        = std::min(requested_size, codec_limit.value_or(requested_size));
+
+  auto const alignment = io::detail::compress_required_chunk_alignment(compression);
+  return std::max(alignment, cudf::util::round_down_safe(size, alignment));
+}
+
+/**
+ * @brief Compresses device data and writes it to the sink.
+ *
+ * Only ZSTD is supported, because decompressing concatenated frames using standard tools yields the
```
Comment (vyasr) [inline cpp/src/io/csv/writer_impl.cu:100]: I know we check this much higher in the stack, but I don't see any checks in any of the write functions in this file, we're dispatching for all compression types. Maybe an extra CUDF_EXPECTS is worthwhile to add? It's cheap and would prevent us from incorrectly using this function in the future. Although maybe some of the helper functions called with `compression` as an arg below will automatically provide that error for us?

**Review by vyasr — APPROVED** [review]
LGTM

File: cpp/src/io/csv/writer_impl.cu (hunk 17 above)
Comment (vuule) [reply cpp/src/io/csv/writer_impl.cu:100]: Valid compression codec is an invariant of the options object. We don't verify these invariants in readers/writers and instead assume that the passed options object is valid.
